### ISSUE TITLE
Improve output of recursive scripts (i.e. pnpm recursive run ...)

### ISSUE THREAD

**Issue by KSXGitHub** [pnpm/pnpm#1440]
This is how pnpm currently displays output of recursive commands:

![Image](https://user-images.githubusercontent.com/11488886/47437662-40cc4400-d7d3-11e8-9f8a-ad76779b23e1.png)

## Problem

Information on the right are obscured, lost.

## Suggestion

<pre>
<b>packages/foo:</b> test$ node ./test.js
  | Maecenas vitae lectus sit ...
  | Maecenas maximus arcu ...
  | Cras sed magna sed ...
  done. (status: 0)

<b>packages/bar:</b> test$ node ./test.js
  | Maecenas vitae lectus sit ...
  | Maecenas maximus arcu ...
  | Cras sed magna sed ...
  done. (status: 0)

</pre>

**Comment by KSXGitHub** [comment]
BTW, I made a sample image of what it should look like:

[![Sample Image](https://user-images.githubusercontent.com/11488886/47513450-7b53e080-d8a8-11e8-8e02-fb05e124523f.png)](https://github.com/KSXGitHub/pnpm-issue-1/blob/3697b0ab7ed260ebe747995e0d51f8c4b73f8a70/sample.js)

**Comment by zkochan** [comment]
seems ok. Though the status on done is redundant because it is always 0

**Comment by KSXGitHub** [comment]
Revamped!

[![Sample](https://github.com/KSXGitHub/pnpm-issue-1/blob/b396b7f5aa5ec1382d5bc177fcaa48b82966f191/sample.png?raw=true)](https://github.com/KSXGitHub/pnpm-issue-1/blob/b396b7f5aa5ec1382d5bc177fcaa48b82966f191/sample.js)

**Comment by KSXGitHub** [comment]
@zkochan 

Hello again, have you touched this issue yet? If not, I have a suggestion for you: I've created a npm package call [`split-shell-buffer`](https://npmjs.com/package/split-shell-buffer) ([source](https://github.com/ksxnodeapps/fun-little-utilities/tree/master/packages/split-shell-buffer) | [test](https://github.com/ksxnodeapps/fun-little-utilities/tree/master/test/split-shell-buffer) | [monorepo](https://github.com/ksxnodeapps/fun-little-utilities)). This package may help you stream data from a child process (stdout + stderr) to the main output (`process.stdout`) with prefix to each line whilst preserving shell colors (even when these color span multiple lines).

**TL;DR** It transforms output of a program from this

![original](https://user-images.githubusercontent.com/11488886/49522377-227d5c00-f8da-11e8-8254-887af7035742.png)

To this

![transformed](https://user-images.githubusercontent.com/11488886/49522545-7daf4e80-f8da-11e8-8ba4-9fd6ab1ff47d.png)

**Comment by zkochan** [comment]
ok, let's try to do it using your lib.

However, I think during installation we just need to hide the output if it is successful. It is how Yarn does it and npm plans to do it the same way.

We can use it for `pnpm m run`, `pnpm m t`, `pnpm m exec`, etc.

**Comment by KSXGitHub** [comment]
> However, I think during installation we just need to hide the output if it is successful. It is how Yarn does it and npm plans to do it the same way.

This library only concern with streaming data, hiding output is out of scope. However, one can iterate the library as an async iterable and count number of lines printed; upon completion, pnpm should be able to print some sort of special control sequence that erases these lines. It looks something like this:

```typescript
const { lines, fromChildProcess } = await import('split-shell-buffer')

let lineCount = 0
const splitter = fromChildProcess(cp).withPrefix(() => Buffer.from('  | '))
const closeEventPromise = new Promise<number>(rs => cp.on('close', x => rs(x)))

for await (const currentLine of lines(splitter)) {
  process.stdout.write(Buffer.from([...currentLine, 10])) // 10 = '\n'.charCodeAt(0)
  lineCount += 1
}

// erasePreviousLines takes a number and returns a Buffer of escape sequence
const status = await closeEventPromise
if (status === 0) process.stdout.write(erasePreviousLines(lineCount))
```

**Comment by KSXGitHub** [comment]
BTW, the current implementation of how pnpm displays subprocess data is flawed: It's somehow unable to print the first meaningful line no matter how hard I try (that is, print some leading `\n`s as a workaround). So, even if clearing lines is not possible, I would still prefer preserving information to a clean terminal.

**Comment by zkochan** [comment]
:ship: 2.23.0-0
